strictclasses/strictclass.py

```diff
@@ -1,7 +1,27 @@
+from functools import partial
+from dataclasses import dataclass
+
+
 def strict(cls):
     def _strict(self):
         for attr, _type in self.__annotations__.items():
             assert isinstance(getattr(self, attr), _type), f'{attr} is not an instance of {_type}'
 
     cls.strict = _strict
     return cls
+
+
+def strict_dataclass(_cls=None, **kwargs):
+    if _cls is None:
+        return partial(strict_dataclass, **kwargs)
+    cls = dataclass(_cls, **kwargs)
+    strict_cls = type(cls.__name__, (_StrictData, cls), {})
+    strict_cls.__module__ = cls.__module__
+    return strict_cls
+
+
+class _StrictData:
+    def __setattr__(self, key, value):
+        _type = self.__annotations__[key]
+        assert isinstance(value, _type), f'{key} is not an instance of {_type}'
+        super().__setattr__(key, value)
```

Comment:
This is very cool! I think that it is a better API than the one I put, maybe we should keep only this one.
About the name, maybe we just use `strictclass`? shorter, but maybe not descriptive enough.
Also, maybe we should allow using it with or without the dataclass decorator. i.e:
```
@strictclass
@dataclass
class Foo:
...
```
and also
```
@strictclass
class Foo:
...
```